### ISSUE TITLE
Akka HTTP 10.0.8 is misnumbered

### ISSUE THREAD

**Issue by sarahgerweck** [akka/akka-http#1241]
As the release notes even say, Akka HTTP 10.0.8 makes a change that breaks backwards compatibility by introducing the new `ContentType.WithMissingCharset` case. This should have triggered a version bump to 10.1 at the very least, and arguably version 11.0.

The release notes even tell people that they need to look out for new compiler warnings: a clear indication that this is a breaking change to the API. This has led to the release of an emergency hotfix in 10.0.9, and I suspect you may have to do another to deal with #1239, which caused my company a lot of pain.

I don't know if Akka HTTP has a written versioning policy, but the most popular scheme by far is [SemVer](http://semver.org/), which mandates a change to the major number if there's a change that is not fully backwards compatible. I'm not saying SemVer is perfect, but it does seem quite unusual to make changes that are known to require downstream code changes in a patch release.

I recognize that perhaps the ship has sailed, but I think you should at least consider putting out a 10.0.10 with the `MissingCharset` system reverted, and putting the `MissingCharset` system back into a release that's numbered 11.0 or 10.1 instead. If you don't want to do this, I'd suggest bumping the number and putting a warning bar on the 10.0.8 and 10.0.9 release notes that they are not backwards compatible.

I appreciate all the hard work on Akka HTTP and I appreciate that Akka HTTP is moving fast and in the right direction. I think it's a fantastic project, but I do hope you will avoid changing APIs in patch releases in the future.

**Comment by raboof** [comment]
First of all, I'm sorry to hear the regression we introduced in 10.0.8 caused you pain - I'm quite confident our fix in 10.0.9 should solve #1239 as well.

As for our versioning scheme: indeed we chose not to adopt semver, we use more flexible rules keeping binary compatibility as laid out in https://github.com/akka/akka-http/blob/master/CONTRIBUTING.md#binary-compatibility . We're quite serious about those, and even developed the MiMa tooling to safeguard against unintentional breakage. We will document that more clearly as part of #1106.

In this particular case, the code that changed was fairly 'deep', and we suspect it is relatively unlikely there is a lot of client code impacted by this change. In a strict SemVer project, perhaps this would have been declared not to be part of the public API and shielded off from the user entirely. We think we strike a reasonable balance between allowing 'power users' to reach into internals, while not being so strict that it would lead to paralysis or ballooning version numbers. For the vast majority of client code this release will be completely compatible - indeed for those rare cases where people matched on this structure anyway we added the warning bar to the 10.0.8 release.

In this case, it appears to me that the fact that we (in semver terms) broke compatibility is not what caused your pain: the fact that we unfortunately caused a regression in our own code did. We strive to improve our process to not allow such things to slip through. In this specific case we don’t think removing the change and retroactively bumping the version will help improve our users situation though, as we expect the population of developers impacted by this is rather small.
